### PR TITLE
[SYCLomatic] [DOC] Add new options and warning in doc folder

### PR DESCRIPTION
**Pull request by zhiweij1**
Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

**Review by zhimingwang36 — APPROVED** [review]
LGTM

File: docs/dev_guide/command-line-options-reference.rst [hunk 1]
```diff
@@ -51,6 +50,8 @@ in alphabetical order.
        header file name. Default: ``dpct``.
    * - ``--enable-ctad``
      - Use a C++17 class template argument deduction (CTAD) in your generated code.
+   * - ``--enable-profiling``
+     - Enables sycl queue profiling in helper functions. Default: off.
```
Comment (krisdale) [inline docs/dev_guide/command-line-options-reference.rst]: Update to "Enable SYCL queue ..."

Formatting for rendered docs: Add double back ticks (``) default value (off).

File: docs/dev_guide/command-line-options-reference.rst [hunk 2]
```diff
@@ -109,10 +110,12 @@ in alphabetical order.
        option will be ignored if the replacement option ``--use-explicit-namespace``
        is used.
    * - ``--no-dpcpp-extensions=<value>``
-     - A comma-separated list of extensions not to be used in migrated code.
-       By default, these extensions will be used in migrated code.
+     - A comma separated list of extensions not to be used in migrated code.
```
Comment (krisdale) [inline docs/dev_guide/command-line-options-reference.rst]: "comma-separated" is correct and already in use in multiple locations in the doc. Lets stay consistent using "comma-separated".

File: docs/dev_guide/command-line-options-reference.rst [hunk 3]
```diff
@@ -225,8 +228,17 @@ in alphabetical order.
        - ``=all``: Generate a complete set of helper header files and place them
          in the ``--out-root`` directory.
    * - ``--use-dpcpp-extensions=<value>``
-     - A comma-separated list of extensions to be used in migrated code. By
-       default, these extensions are not used in migrated code.
+     - A comma separated list of extensions to be used in migrated code.
```
Comment (krisdale) [inline docs/dev_guide/command-line-options-reference.rst]: "comma-separated"

File: docs/dev_guide/command-line-options-reference.rst [hunk 4]
```diff
@@ -109,10 +110,12 @@ in alphabetical order.
        option will be ignored if the replacement option ``--use-explicit-namespace``
        is used.
    * - ``--no-dpcpp-extensions=<value>``
-     - A comma-separated list of extensions not to be used in migrated code.
-       By default, these extensions will be used in migrated code.
+     - A comma separated list of extensions not to be used in migrated code.
+       By default, these extensions are used in migrated code.
 
        - ``=enqueued_barriers``: Enqueued barriers extension.
+       - ``=device_info``: Use Intel's Extensions for Device Information if it is
```
Comment (krisdale) [inline docs/dev_guide/command-line-options-reference.rst]: Update to "Use the Intel extensions for device information if supported ..."

File: docs/dev_guide/command-line-options-reference.rst [hunk 5]
```diff
@@ -225,8 +228,17 @@ in alphabetical order.
        - ``=all``: Generate a complete set of helper header files and place them
          in the ``--out-root`` directory.
    * - ``--use-dpcpp-extensions=<value>``
-     - A comma-separated list of extensions to be used in migrated code. By
-       default, these extensions are not used in migrated code.
+     - A comma separated list of extensions to be used in migrated code.
+       By default, these extensions are not used in migrated code.
+
+       - ``=c_cxx_standard_library``: Use std functions from libdevice (provided by
```
Comment (krisdale) [inline docs/dev_guide/command-line-options-reference.rst]: Update to "... from the libdevice library ..."

File: docs/dev_guide/command-line-options-reference.rst [hunk 6]
```diff
@@ -225,8 +228,17 @@ in alphabetical order.
...
          in the ``--out-root`` directory.
    * - ``--use-dpcpp-extensions=<value>``
-     - A comma-separated list of extensions to be used in migrated code. By
-       default, these extensions are not used in migrated code.
+     - A comma separated list of extensions to be used in migrated code.
+       By default, these extensions are not used in migrated code.
+
+       - ``=c_cxx_standard_library``: Use std functions from libdevice (provided by
+         Intel(R) oneAPI DPC++/C++ Compiler) and C/C++ standard library to migrate
+         functions which has no mapping in SYCL standard.
```
Comment (krisdale) [inline docs/dev_guide/command-line-options-reference.rst]: update to "... which have no mapping in the SYCL..."

File: docs/dev_guide/command-line-options-reference.rst [hunk 7]
```diff
@@ -225,8 +228,17 @@ in alphabetical order.
...
+         If this value is used together with intel_device_math, intel_device_math
```
Comment (krisdale) [inline docs/dev_guide/command-line-options-reference.rst]: Update to "If this value is used together with intel_device_math, the intel_device_math ..."

Formatting for rendered docs: Add double back ticks (``) around function names/namespace.

File: docs/dev_guide/command-line-options-reference.rst [hunk 8]
```diff
@@ -225,8 +228,17 @@ in alphabetical order.
...
+         functions take precedence.
+       - ``=intel_device_math``: Use sycl::ext::intel::math functions from libdevice
```
Comment (krisdale) [inline docs/dev_guide/command-line-options-reference.rst]: Update to " ... from the libdevice library .."

Formatting for rendered docs: Add double back ticks (``) around function names/namespace.

File: docs/dev_guide/command-line-options-reference.rst [hunk 9]
```diff
@@ -225,8 +228,17 @@ in alphabetical order.
...
+         (provided by Intel(R) oneAPI DPC++/C++ Compiler) to migrate functions which
+         have no mapping in SYCL standard.
```
Comment (krisdale) [inline docs/dev_guide/command-line-options-reference.rst]: "... in the SYCL .."

File: docs/dev_guide/diagnostic_ref/dpct1103.rst [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+DPCT1103
+========
+
+Message
+-------
+
+.. _msg-1103-start:
+
+``<expression text>`` should be a dynamic library. The dynamic library should supply
+"wrapped" kernel functions.
```
Comment (krisdale) [inline docs/dev_guide/diagnostic_ref/dpct1103.rst]: Update to "... supply wrapped kernel ..."

File: docs/dev_guide/diagnostic_ref/dpct1104.rst [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+DPCT1104
+========
+
+Message
+-------
+
+.. _msg-1104-start:
+
+``<expression text>`` should point to a dynamic library loaded in memory. The dynamic
+library should supply "wrapped" kernel functions.
```
Comment (krisdale) [inline docs/dev_guide/diagnostic_ref/dpct1104.rst]: Update to "... supply wrapped kernel ..."

File: docs/dev_guide/diagnostic_ref/dpct1105.rst [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+DPCT1105
+========
+
+Message
+-------
+
+.. _msg-1105-start:
+
+``<expression text>`` should point to a dynamic library loaded in memory. The dynamic
+library should supply "wrapped" kernel functions. Compilation options passed to
```
Comment (krisdale) [inline docs/dev_guide/diagnostic_ref/dpct1105.rst]: Update to "... supply wrapped kernel ..."

File: docs/dev_guide/diagnostic_ref/dpct1106.rst [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+
+DPCT1106
+========
+
+Message
+-------
+
+.. _msg-1106-start:
+
+``<function name>`` was migrated with Intel's Extensions for Device Information which
```
Comment (krisdale) [inline docs/dev_guide/diagnostic_ref/dpct1106.rst]: Update to "... with the Intel extensions for device information ..."

File: docs/dev_guide/command-line-options-reference.rst [hunk 14]
```diff
@@ -225,8 +228,17 @@ in alphabetical order.
        - ``=all``: Generate a complete set of helper header files and place them
          in the ``--out-root`` directory.
    * - ``--use-dpcpp-extensions=<value>``
-     - A comma-separated list of extensions to be used in migrated code. By
-       default, these extensions are not used in migrated code.
+     - A comma separated list of extensions to be used in migrated code.
+       By default, these extensions are not used in migrated code.
+
+       - ``=c_cxx_standard_library``: Use std functions from libdevice (provided by
+         Intel(R) oneAPI DPC++/C++ Compiler) and C/C++ standard library to migrate
```
Comment (krisdale) [inline docs/dev_guide/command-line-options-reference.rst]: "C/C++ Standard Library"

File: docs/dev_guide/command-line-options-reference.rst [hunk 15]
```diff
@@ -35,9 +35,8 @@ in alphabetical order.
    * - ``--build-script-file=<file>``
      - Specifies the name of generated makefile for migrated file(s). Default
        name: ``Makefile.dpct``.
-   * - ``=c_cxx_standard_library``
-     - A set of functions from the C and C++ standard libraries that are allowed
-       to be used in SYCL device code.
+   * - ``--change-cuda-files-extension-only``
+     - Limit extension change to .cu and .cuh files only. Default: off.
```
Comment (krisdale) [inline docs/dev_guide/command-line-options-reference.rst]: Formatting for rendered docs: Add double back ticks (``) around file extensions (.cu, .cuh) and default value (off).

File: docs/dev_guide/diagnostic_ref/dpct1105.rst [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+cuModuleLoadDataEx are not migrated.
```
Comment (krisdale) [inline docs/dev_guide/diagnostic_ref/dpct1105.rst]: Formatting for rendered docs: Add double back ticks (``) around function names.

**Review by krisdale — CHANGES_REQUESTED** [review]
Editorial review completed for existing content. 

Before merging, dpct message detail pages need to be completed (addition of Detailed Help and Suggestions to Fix sections).

File: docs/dev_guide/diagnostic_ref/dpct1101.rst [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+
+``<expression text>`` expression was replaced with a value. Modify the code to
+use the original expression, provided in comments, if it is correct.
+
+.. _msg-1101-end:
+
+Detailed Help
+-------------
+
+Identities including constant variable name(s), type(s) and macro(s) used in the
```
Comment (krisdale) [inline docs/dev_guide/diagnostic_ref/dpct1101.rst]: Small edit: add comma after "type(s)"

File: docs/dev_guide/diagnostic_ref/dpct1101.rst [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+migrated SYCL code are replaced with their value since they may not be available
+in the code scope. User can replace the value with the original expression listed
+in the comment manually after confirming that the identities are available in the
+code scope.
+
+Suggestions to Fix
+------------------
+
+Use macro or const variable to replace the value.
```
Comment (krisdale) [inline docs/dev_guide/diagnostic_ref/dpct1101.rst]: "Use **a** macro or ..."

File: docs/dev_guide/diagnostic_ref/dpct1101.rst [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+
+Example:
+
+Original code:
```
Comment (krisdale) [inline docs/dev_guide/diagnostic_ref/dpct1101.rst]: Replace with: "For example, this original CUDA* code:

File: docs/dev_guide/diagnostic_ref/dpct1101.rst [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+   :linenos:
+   __global__ void f() {
+     const int x = 2;
+     __shared__ int fold[x];
+   }
+   int main() {
+     f<<<1, 1>>>();
+     return 0;
+   }
+Migrated code:
```
Comment (krisdale) [inline docs/dev_guide/diagnostic_ref/dpct1101.rst]: Replace with: "results in the following migrated SYCL code:"

(And make sure there is a return line between the code block and text)

File: docs/dev_guide/diagnostic_ref/dpct1101.rst [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+   
+       cgh.parallel_for(
+           sycl::nd_range<3>(sycl::range<3>(1, 1, 1), sycl::range<3>(1, 1, 1)),
+           [=](sycl::nd_item<3> item_ct1) {
+             f(fold_acc_ct1.get_pointer());
+           });
+     });
+     return 0;
+   }
+Fixed code:
```
Comment (krisdale) [inline docs/dev_guide/diagnostic_ref/dpct1101.rst]: Replace with : "which is rewritten to:"

File: docs/dev_guide/diagnostic_ref/dpct1102.rst [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+.. _msg-1102-start:
+
+Zero-length arrays are not permitted in SYCL device code.
+
+.. _msg-1102-end:
+
+Detailed Help
+-------------
+
+Zero-length arrays are not supported in C++ standard and SYCL Spec. It may be
```
Comment (krisdale) [inline docs/dev_guide/diagnostic_ref/dpct1102.rst]: "... in the C++ standard or SYCL specification."

File: docs/dev_guide/diagnostic_ref/dpct1102.rst [hunk 23]
```diff
@@ -0,0 +1,91 @@
...
+supported by some host compiler implementations, but it is not supported by
+Intel(R) oneAPI DPC++/C++ Compiler for device code.
```
Comment (krisdale) [inline docs/dev_guide/diagnostic_ref/dpct1102.rst]: Use "®"

File: docs/dev_guide/diagnostic_ref/dpct1102.rst [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+
+Suggestions to Fix
+------------------
+
+Use array which length is greater than zero.
```
Comment (krisdale) [inline docs/dev_guide/diagnostic_ref/dpct1102.rst]: "Use an array with length greater than..."

File: docs/dev_guide/diagnostic_ref/dpct1102.rst [hunk 25]
```diff
@@ -0,0 +1,91 @@
...
+
+Example:
```
Comment (krisdale) [inline docs/dev_guide/diagnostic_ref/dpct1102.rst]: Follow the same pattern with the example code as I showed in dpct1101. 

"For example, this original CUDA* code:"
[original code block]

"results in the following migrated SYCL code:"
[migrated code block]

"which is rewritten to:"
[rewritten code block]

This pattern is already in place, as seen in the existing doc for DPCT1086.

File: docs/dev_guide/diagnostic_ref/dpct1103.rst [hunk 26]
```diff
@@ -0,0 +1,72 @@
...
+
+``<expression text>`` should be a dynamic library. The dynamic library should supply
+wrapped kernel functions.
+
+.. _msg-1103-end:
+
+Detailed Help
+-------------
+
+The CUDA application expects ``<expression text>`` to be the name of a CUDA module.
```
Comment (krisdale) [inline docs/dev_guide/diagnostic_ref/dpct1103.rst]: "The CUDA*..."

File: docs/dev_guide/diagnostic_ref/dpct1103.rst [hunk 27]
```diff
@@ -0,0 +1,72 @@
...
+The user should ensure that the CUDA code that generated the CUDA module is migrated
+to SYCL code and compiled to generate a kernel library.
+
+Suggestions to Fix
+------------------
+
+The user should find the module source file(s) that are used to make the CUDA module
+and migrate them to SYCL with the --extra-arg=-ptx option of the compatibility tool
```
Comment (krisdale) [inline docs/dev_guide/diagnostic_ref/dpct1103.rst]: "... migrate them to SYCL using the .."

and wrap the option in double back ticks (``)

Replace "compatibility tool" with "|tool_name|"

File: docs/dev_guide/diagnostic_ref/dpct1103.rst [hunk 28]
```diff
@@ -0,0 +1,72 @@
...
+
+.. code-block:: bash
+   :linenos:
+   icpx -fsycl <migrated module source file(s)> -shared -o <kernel library filename>
+
+The ``<kernel library filename>`` should be chosen so that ``<expression text>`` will have
+the same filename at runtime. For the following example, ``<kernel library filename>`` should
+be set to ``library.so``.
+
+Example:
```
Comment (krisdale) [inline docs/dev_guide/diagnostic_ref/dpct1103.rst]: Please apply the same pattern as I described in dpct1101 and dpct1102

File: docs/dev_guide/diagnostic_ref/dpct1104.rst [hunk 29]
```diff
@@ -0,0 +1,73 @@
...
+
+``<expression text>`` should point to a dynamic library loaded in memory. The dynamic
+library should supply wrapped kernel functions.
+
+.. _msg-1104-end:
+
+Detailed Help
+-------------
+
+The CUDA application expects ``<expression text>`` to point to a CUDA module whose file
```
Comment (krisdale) [inline docs/dev_guide/diagnostic_ref/dpct1104.rst]: "The CUDA* ..."

File: docs/dev_guide/diagnostic_ref/dpct1104.rst [hunk 30]
```diff
@@ -0,0 +1,73 @@
...
+image has been copied to memory. The user should ensure that the CUDA code that generated
+the CUDA module is migrated to SYCL code, compiled to generate a kernel library, and then
+copied to memory by the migrated SYCL application.
+
+Suggestions to Fix
+------------------
+
+The user should find the module source file(s) that are used to make the CUDA module and
+migrate the module source file(s) to SYCL with the --extra-arg=--ptx option of the
```
Comment (krisdale) [inline docs/dev_guide/diagnostic_ref/dpct1104.rst]: "... to SYCL using the ..."

Wrap option in double back ticks (``)

Replace "compatibility tool" with "|tool_name|"

File: docs/dev_guide/diagnostic_ref/dpct1104.rst [hunk 31]
```diff
@@ -0,0 +1,73 @@
...
+
+.. code-block:: bash
+   :linenos:
+   icpx -fsycl <migrated module source file(s)> -shared -o <kernel library filename>
+
+The user should ensure that the application will load the kernel library image into
+memory at runtime. For the following example this can be ensured by setting
+``<kernel library filename>`` to ``library.so``.
+
+Example:
```
Comment (krisdale) [inline docs/dev_guide/diagnostic_ref/dpct1104.rst]: Please apply the same pattern as described in dpct1101.

File: docs/dev_guide/diagnostic_ref/dpct1106.rst [hunk 32]
```diff
@@ -0,0 +1,57 @@
...
+may be not supported by all compilers or runtimes. You may need to adjust the code.
+
+.. _msg-1106-end:
+
+Detailed Help
+-------------
+
+The CUDA ``<function name>`` for example ``cudaMemGetInfo`` or ``cuMemGetInfo`` was
+migrated with Intel's Extensions for Device Information, which may be not supported
+by all compilers or runtimes.
```
Comment (krisdale) [inline docs/dev_guide/diagnostic_ref/dpct1106.rst]: "The **CUDA*** **function** ``<function name>`` (for example ``cudaMemGetInfo`` or ``cuMemGetInfo``) was
 migrated with **the** **Intel extensions for device information**, which may be not supported
 by all compilers or runtimes. **You may need to adjust the code.**"

File: docs/dev_guide/diagnostic_ref/dpct1106.rst [hunk 33]
```diff
@@ -0,0 +1,57 @@
...
+
+Original code:
```
Comment (krisdale) [inline docs/dev_guide/diagnostic_ref/dpct1106.rst]: Please apply the same pattern as described in dpct1101.

**Review by krisdale — CHANGES_REQUESTED** [review]
Overall looks good. I made some suggestions for both grammar and to be consistent with patterns in use in existing docs.  In general, make sure to have a line between the end of a code block and the following text (Sphinx is sensitive to whitespace and lines between content blocks). 

If you apply these changes, I approve.

File: docs/dev_guide/diagnostic_ref/dpct1106.rst [hunk 34]
```diff
@@ -0,0 +1,57 @@
...
+DPCT1106
+========
+
+Message
+-------
+
+.. _msg-1106-start:
+
+``<function name>`` was migrated with the Intel extensions for device information which
+may be not supported by all compilers or runtimes. You may need to adjust the code.
```
Comment (krisdale) [inline docs/dev_guide/diagnostic_ref/dpct1106.rst]: "... may not be supported ..."

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**